pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -0,0 +1,344 @@
+	"fmt"
+	"github.com/openshift/library-go/pkg/crypto"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"strings"
+	"time"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/clustermembercontroller"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	corev1client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const (
+	workQueueKey     = "key"
+	EtcdCertValidity = 3 * 365 * 24 * time.Hour
+	caNamespace      = "openshift-config"
+	etcdNamespace    = "openshift-etcd"
+)
+
+type EtcdCertSignerController struct {
+	clientset corev1client.Interface
+	// Not using this but still keeping it in there
+	operatorConfigClient v1helpers.OperatorClient
+	queue                workqueue.RateLimitingInterface
+	eventRecorder        events.Recorder
+}
+
+func NewEtcdCertSignerController(
+	clientset corev1client.Interface,
+	operatorConfigClient v1helpers.OperatorClient,
+
+	kubeInformersForOpenshiftEtcdNamespace informers.SharedInformerFactory,
+	eventRecorder events.Recorder,
+) *EtcdCertSignerController {
+	c := &EtcdCertSignerController{
+		clientset:            clientset,
+		operatorConfigClient: operatorConfigClient,
+		queue:                workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "EtcdCertSignerController"),
+		eventRecorder:        eventRecorder.WithComponentSuffix("etcd-cert-signer-controller"),
+	}
+	kubeInformersForOpenshiftEtcdNamespace.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+	return c
+}
+
+func (c *EtcdCertSignerController) Run(i int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting ClusterMemberController")
+	defer klog.Infof("Shutting down ClusterMemberController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *EtcdCertSignerController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *EtcdCertSignerController) processNextWorkItem() bool {
+	dsKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(dsKey)
+
+	err := c.sync()
+	if err == nil {
+		c.queue.Forget(dsKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", dsKey, err))
+	c.queue.AddRateLimited(dsKey)
+
+	return true
+}
+
+func (c *EtcdCertSignerController) sync() error {
+	// TODO: make the namespace and name constants in one of the packages
+	cm, err := c.clientset.CoreV1().ConfigMaps(etcdNamespace).Get("member-config", metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("error getting configmap %#v\n", err)
+		return err
+	}
+	scaling := &clustermembercontroller.EtcdScaling{}
```

Comment:
Just a thought and I need more context. What if we just used the existence of the configmap as an action item but then actually populated the data from the observation controller vs data from configmap.

Reply:
Basically do we want the configmap to be the source of truth? Can we achieve the same logic only using observations?

Reply:
Feels safer to not rely on data that admin can hack for certs. If someone hacks configmap for adding peer they won't ever get added because the certs won't be valid. Lets just add that as a TODO